File: app/src/commands/player.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import type { Option } from '@tb-dev/utils';
import { invoke } from '@tauri-apps/api/core';
import type {
  GetPlayerCoordsRequest,
  GetPlayerCoordsResponse,
  GetPlayerIdsRequest,
  GetPlayerIdsResponse,
  GetPlayerMaintenanceRequest,
  GetPlayerMaintenanceResponse,
  GetPlayerMilitaryRequest,
  GetPlayerMilitaryResponse,
  GetPlayerReportsRequest,
  GetPlayerReportsResponse,
  GetPlayerRequest,
  GetPlayerResponse,
  GetPlayerStatusRequest,
  GetPlayerStatusResponse,
  GetPlayerStorageCapacityRequest,
  GetPlayerStorageCapacityResponse,
  GetPlayerWorldsRequest,
  GetPlayerWorldsResponse,
  GetPublicPlayerRequest,
  GetPublicPlayerResponse,
  GetPublicPlayersRequest,
  GetPublicPlayersResponse,
  PlayerExistsRequest,
  PlayerExistsResponse,
  PlayerId,
  PlayerOptions,
  PlayerStatus,
  SetPlayerStatusRequest,
  SpawnPlayerRequest,
} from '@/types/bindings';

export async function getPlayer() {
  const req: GetPlayerRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetPlayerResponse>('get_player', { req });
}

export async function getPlayerCoords(id: PlayerId) {
  const req: GetPlayerCoordsRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  return invoke<GetPlayerCoordsResponse>('get_player_coords', { req });
}

export async function getPlayerIds() {
  const req: GetPlayerIdsRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetPlayerIdsResponse>('get_player_ids', { req });
}

export async function getPlayerMaintenance() {
  const req: GetPlayerMaintenanceRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetPlayerMaintenanceResponse>('get_player_maintenance', { req });
}

export async function getPlayerMilitary() {
  const req: GetPlayerMilitaryRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetPlayerMilitaryResponse>('get_player_military', { req });
}

export async function getPlayerReports() {
  const req: GetPlayerReportsRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetPlayerReportsResponse>('get_player_reports', { req });
}

export async function getPlayerStatus(id: PlayerId) {
  const req: GetPlayerStatusRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  return invoke<GetPlayerStatusResponse>('get_player_status', { req });
}

export async function getPlayerStorageCapacity() {
  const req: GetPlayerStorageCapacityRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetPlayerStorageCapacityResponse>('get_player_storage_capacity', { req });
}

export async function getPlayerWorlds(id: PlayerId) {
  const req: GetPlayerWorldsRequest = { id };
  return invoke<GetPlayerWorldsResponse>('get_player_worlds', { req });
}

export async function getPublicPlayer(id: PlayerId) {
  const req: GetPublicPlayerRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  return invoke<GetPublicPlayerResponse>('get_public_player', { req });
}

export async function getPublicPlayers() {
  const req: GetPublicPlayersRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetPublicPlayersResponse>('get_public_players', { req });
}

export async function playerExists(id: PlayerId) {
  const req: PlayerExistsRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  return invoke<PlayerExistsResponse>('player_exists', { req });
}

export async function setPlayerStatus(status: PlayerStatus) {
  const req: SetPlayerStatusRequest = {
    world: NIL.world.getIdStrict(),
    status,
  };

  await invoke('set_player_status', { req });
}

export async function spawnPlayer(options: PlayerOptions, worldPassword: Option<string>) {
  const req: SpawnPlayerRequest = {
    world: NIL.world.getIdStrict(),
    worldPassword: worldPassword ?? null,
    options,
  };

  await invoke('spawn_player', { req });
}

File: app/src-tauri/src/command/player.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::manager::ManagerExt;
use nil_payload::request::player::*;
use nil_payload::response::player::*;
use tauri::AppHandle;

#[tauri::command]
pub async fn get_player(app: AppHandle, req: GetPlayerRequest) -> Result<GetPlayerResponse> {
  app
    .client(async |cl| cl.get_player(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_coords(
  app: AppHandle,
  req: GetPlayerCoordsRequest,
) -> Result<GetPlayerCoordsResponse> {
  app
    .client(async |cl| cl.get_player_coords(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_ids(
  app: AppHandle,
  req: GetPlayerIdsRequest,
) -> Result<GetPlayerIdsResponse> {
  app
    .client(async |cl| cl.get_player_ids(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_maintenance(
  app: AppHandle,
  req: GetPlayerMaintenanceRequest,
) -> Result<GetPlayerMaintenanceResponse> {
  app
    .client(async |cl| cl.get_player_maintenance(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_military(
  app: AppHandle,
  req: GetPlayerMilitaryRequest,
) -> Result<GetPlayerMilitaryResponse> {
  app
    .client(async |cl| cl.get_player_military(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_status(
  app: AppHandle,
  req: GetPlayerStatusRequest,
) -> Result<GetPlayerStatusResponse> {
  app
    .client(async |cl| cl.get_player_status(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_storage_capacity(
  app: AppHandle,
  req: GetPlayerStorageCapacityRequest,
) -> Result<GetPlayerStorageCapacityResponse> {
  app
    .client(async |cl| cl.get_player_storage_capacity(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_worlds(
  app: AppHandle,
  req: GetPlayerWorldsRequest,
) -> Result<GetPlayerWorldsResponse> {
  app
    .client(async |cl| cl.get_player_worlds(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_public_player(
  app: AppHandle,
  req: GetPublicPlayerRequest,
) -> Result<GetPublicPlayerResponse> {
  app
    .client(async |cl| cl.get_public_player(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_public_players(
  app: AppHandle,
  req: GetPublicPlayersRequest,
) -> Result<GetPublicPlayersResponse> {
  app
    .client(async |cl| cl.get_public_players(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn player_exists(
  app: AppHandle,
  req: PlayerExistsRequest,
) -> Result<PlayerExistsResponse> {
  app
    .client(async |cl| cl.player_exists(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn set_player_status(app: AppHandle, req: SetPlayerStatusRequest) -> Result<()> {
  app
    .client(async |cl| cl.set_player_status(req).await)
    .await
    .map_err(Into::into)
}

#[tauri::command]
pub async fn spawn_player(app: AppHandle, req: SpawnPlayerRequest) -> Result<()> {
  app
    .client(async |cl| cl.spawn_player(req).await)
    .await
    .map_err(Into::into)
}
